src/Type/Php/RegexArrayShapeMatcher.php

```diff
@@ -653,15 +656,22 @@ private function walkGroupAst(TreeNode $ast, TrinaryLogic &$isNonEmpty, TrinaryL
 
 	private function getLiteralValue(TreeNode $node): ?string
 	{
-		if ($node->getId() === 'token' && $node->getValueToken() === 'literal') {
+		if ($node->getId() === 'token' && in_array($node->getValueToken(), ['literal', 'class:literal'], true)) {
 			return $node->getValueValue();
 		}
 
-		// literal "-" outside of a character class like '~^((\\d{1,6})-)$~'
-		if ($node->getId() === 'token' && $node->getValueToken() === 'range') {
+		// literal "-" in front/back of a character class like '[-a-z]' or '[abc-]', not forming a range
+		if ($node->getId() === 'token' && $node->getValueToken() === 'class:range') {
 			return $node->getValueValue();
 		}
 
+		// [:digit:] and the like, more support coming later
+		if ($node->getId() === 'token' && $node->getValueToken() === 'class:posix_class') {
+			if ($node->getValueValue() === '[:digit:]') {
+				return implode('', range(0, 9));
+			}
+		}
```

Comment:
@staabm I think it's mergeable as is now, but just for future reference, there is a lot more we can support here:

In terms of token types, you can use this to get a full list of tokens that can appear within character classes:

```
<?php
require 'vendor/autoload.php';

$regex = '{[-\p{L}[\]*|\x03\a\b+?{}(?:)-]+[^[:digit:]?{}a-z0-9k-#]+a-z}';
$grammar  = new Hoa\File\Read('resources/RegexGrammar.pp');
$compiler = Hoa\Compiler\Llk\Llk::load($grammar);
$ast      = $compiler->parse($regex);
$dump     = new Hoa\Compiler\Visitor\Dump();
echo $dump->visit($ast);
```

Output with some notes:

```
>  >  >  #quantification
>  >  >  >  #class
>  >  >  >  >  token(class:range, -)
>  >  >  >  >  token(class:character_type, \p{L})
>  >  >  >  >  token(class:class_, [)
>  >  >  >  >  token(class:escaped_end_class, \]) // this is matching a literal `]`, the backslash has to be dropped
>  >  >  >  >  token(class:literal, *)
>  >  >  >  >  token(class:literal, |)
>  >  >  >  >  token(class:dynamic_character, \x03)
>  >  >  >  >  token(class:character, \a)
>  >  >  >  >  token(class:character, \b)
>  >  >  >  >  token(class:literal, +)
>  >  >  >  >  token(class:literal, ?)
>  >  >  >  >  token(class:literal, {)
>  >  >  >  >  token(class:literal, })
>  >  >  >  >  token(class:literal, ()
>  >  >  >  >  token(class:literal, ?)
>  >  >  >  >  token(class:literal, :)
>  >  >  >  >  token(class:literal, ))
>  >  >  >  >  token(class:range, -)
>  >  >  >  token(one_or_more, +)
>  >  >  #quantification
>  >  >  >  #negativeclass
>  >  >  >  >  token(class:posix_class, [:digit:]) // I added tentative support for digit, I can add more later but first want to discuss how to best do this
>  >  >  >  >  token(class:literal, ?)
>  >  >  >  >  token(class:literal, {)
>  >  >  >  >  token(class:literal, })
>  >  >  >  >  #range
>  >  >  >  >  >  token(class:literal, a)
>  >  >  >  >  >  token(class:literal, z)
>  >  >  >  >  #range
>  >  >  >  >  >  token(class:literal, 0)
>  >  >  >  >  >  token(class:literal, 9)
>  >  >  >  >  #range
>  >  >  >  >  >  token(class:literal, k)
>  >  >  >  >  >  token(class:literal, #) // gotta watch out ranges can be from/to any character, this matches any of kjihgfedcba`_^]\[ZYXWVUTSRQPONMLKJIHGFEDCBA@?>=<;:9876543210/.-,+*)('&%$#
>  >  >  >  token(one_or_more, +)
>  >  >  token(literal, a)
>  >  >  token(literal, -)
>  >  >  token(literal, z)
```

Reply:
Thanks for that. If you have more concrete examples of regex patterns which go wrong I am all ears. 

I don't need/want to support any super edge case format, but we should at least try hard to fix any failure in type-inference (e.g where the current logic decides its a numeric-string but its actually just string)

Reply:
Yes I think this could be merged already. Once merged I'll check again how it looks on the Composer repo (I somehow cannot run phpstan from source on the composer project), but I believe this will fix all regex parse errors I encountered. I am not sure though as the parse errors are swallowed.. that's why I wanted a way to surface them to users so they know that phpstan couldn't parse a given regex, and we could get reports and fix those. Right now you only realize it couldn't parse it when the type of $matches is array<string> and that causes further problems in the code.

And yeah I agree supporting everything including constant strings etc may not make sense, but right now I think you probably can end up with patterns where it infers numeric-string wrongly. I'll give that a closer look some other time :)